\section{Clase GameManager}

{\small
\begin{tabular}{r|l}
Nombre del \textit{tester} & Jorge Colao Adán \\
& Daniel León Romero\\
Fecha de asignación & 18 de febrero de 2011 \\
Fecha de finalización & 21 de febrero de 2011 \\
Código bajo prueba & \texttt{GameManager}
\end{tabular}
}

En este apartado se hablará sobre las pruebas realizadas con la herramienta \textit{JUnit} sobre esta clase.

La estrategia a seguir para las pruebas de estos apartados será \textit{Each Choice}. La elección de los valores de los parámetos, se realizarán con valores límite para los atributos númericos y valores propensos a error para el resto.

\subsection{GameManager::updateGameList}

En este primer método como no tiene ningún parámetro lo que tenemos que comprobar es que al hacer el test con \textit{JUnit} no salte ninguna excepción. Además, comprobamos que inicialmente el número de partidas a las que se está unido y las partidas disponibles para unirse es igual a cero, y que después de la ejecución de este caso de uso es distinto de cero.

Debido a que con la herramienta \textit{JUnit} no podemos alcanzar una covertura alta, lo que haremos será hacer pruebas exploratorias.

\subsection{GameManager::createGame}

Este método tiene seis parámetros de entrada, por lo que tenemos que utilizar alguna estrategia de generación de casos de prueba. Hemos utilizado \textit{Each Choice} con la herramienta de la página \url{http://161.67.140.42/CombTestWeb/}. Pero comprobando las combinaciones que hemos obtenido nos hemos dado cuenta de que eran poco exahustivas (Test 1 a 3). Debido a este motivo, hemos añadido otros casos de prueba interesantes (Test 4 a 12), como son; probar para que falle cada parámetro independientemente.

Para elegir tanto los valores de prueba como los casos de pruebas de los casos interesantes se ha analizado el comportamiento de este método (mediante caja blanca). Después de analizar el método nos hemos dado cuenta de que pude haber los siguientes errores:
\begin{itemize}
\item NullPointerException
\item EmptyStringException
\item NegativeValueException
\end{itemize}
Esta comprobación es local, por ello si se lanza alguna excepción de las anteriores, no se creará la partida en el servidor.

\begin{itemize}
\item \textbf{\texttt{name}}
\subitem \textit{null}
\subitem Cadena vacía
\subitem Cadena válida: \texttt{"partida"}

\item \textbf{\texttt{description}}
\subitem \textit{null}
\subitem Cadena vacía
\subitem Cadena válida: \texttt{"partida guerra mundo"}

\item \textbf{\texttt{gameSession}}
\subitem \textit{null}
\subitem Fecha: \texttt{"Hora actual"} y \texttt{"01/Marzo/2012 a las 14:00"}

\item \textbf{\texttt{turnTime}}
\subitem Número incorrecto: \texttt{"0"}
\subitem Número correcto. \texttt{"1"} y \texttt{"112"}

\item \textbf{\texttt{defTime}}
\subitem Número incorrecto: \texttt{"0"}
\subitem Número correcto. \texttt{"1"} y \texttt{"20"}

\item \textbf{\texttt{negTime}}
\subitem Número incorrecto: \texttt{"0"}
\subitem Número correcto. \texttt{"1"} y \texttt{"33"}
\end{itemize}

La única comprobación especial que podemos hacer es mirar si salta alguna excepción, si no salta ninguna excepción el método estará correcto, lo que creará una partida en el servidor.

Esta es la tabla de casos de prueba y resultados esperados.

{\footnotesize
\begin{longtable}[c]{lccc}
 & \textbf{Valores de Prueba} & \textbf{Objetivo del test} & \textbf{Resultado esperado} \\
\hline \hline
\endhead

Test1 & ("partida", "partida guerra mundo",  & crear partida & Funcinamiento Correcto \\
 & Fecha\_Posterior, 112, 1, 1) & & \\
Test2 & (null, null, Fecha\_Actual, & name, description & NullPointer\\
 & 1, 20, 33) & & \\
Test3 & (vacío, vacío, & gameSession & NullPointer\\
 & null, 0, 0, 0) & & \\


Test4 & (vacío, vacío,  & name & EmptyString\\
 &  Fecha\_Posterior, 1, 1, 1) & & \\
Test5 & (null, "partida guerra mundo",  & name & NullPointer\\
 &   Fecha\_Posterior, 1, 1, 1) & & \\
Test6 & ("partida", null,  & description  & NullPointer\\
 &  Fecha\_Actual, 1, 1, 33) & & \\
Test7 & ("partida", "partida guerra mundo",  & gameSession & NullPointer\\
 &   null, 1, 1, 33) & & \\
Test8 & ("partida", "partida guerra mundo",  & turnTime & NegativeValue\\
 &   Fecha\_Actual, 0, 20, 33) & & \\
Test9 & ("partida", "partida guerra mundo",  & defTime & NegativeValue\\
 &   Fecha\_Actual, 112, 0, 33) & & \\
Test10 & ("partida", "partida guerra mundo",  & negTime & NegativeValue\\
 &  Fecha\_Actual, 112, 20, 0) & & \\
Test11 & ("partida", "partida guerra mundo",  & crear partida & Funcinamiento Correcto \\
 &  Fecha\_Actual, 112, 20, 33) & & \\
Test12 & ("partida", vacío,   & crear partida & Funcinamiento Correcto \\
 &  Fecha\_Posterior, 1, 1, 1) & & \\

\hline
\end{longtable}
}

\subsection{GameManager::joinGame}

Para este caso de uso tan sólo tenemos un parámetro que es el número del juego al que queremos unirnos. Por este motivo tenemos que comprobar los valores límite de la lista de partidas a unirse. Para representar los valores límite, hemos usado de límite inferior los números ''--1'' y "0", ya que la comparación es que sea menor de "0". Y para el límite superior usaremos el tope de partidas disponibles en el servidor y el tope más uno. En nuestro servidor tenemos dos partidas a las que nos podemos unir, por lo tanto, el tope sería ``1'' y el tope más uno sería ``2''. 

\begin{itemize}
\item \textbf{\texttt{gameSelected}}
\subitem Límite inferior: \texttt{"\--1"} y \texttt{"0"}
\subitem Límite superior: \texttt{"tope = 1"} y \texttt{"tope + 1 = 2"}
\end{itemize}

También tenemos que comprobar que antes de unirnos a la partida tiene que haber al menos una partida a la que podamos unirnos. Y después de unirnos comprobar que el número de filas en la lista de partidas actuales es mayor que antes y la lista de partidas para unirme es menor.
Las posobles excepciones que puede tener este método son:
\begin{itemize}
\item ArrayIndexOutOfBoundsException, para números por debajo del rango.
\item IndexOutOfBoundsException, pra números por encima del rango.
\end{itemize}

Esta es la tabla de casos de prueba y resultados esperados.

{\footnotesize
\begin{longtable}[c]{lccc}
 & \textbf{Valores de Prueba} & \textbf{Objetivo del test} & \textbf{Resultado esperado} \\
\hline \hline
\endhead

Test1 & (-1) & gameIndex (límite inferior) & ArrayIndexOutOfBounds\\
Test2 & (1) & seleccionar segunda partida & Funcionamiento correcto\\
Test3 & (0) & seleccionar primera partida & Funcionamiento correcto\\
Test4 & (2) & gameIndex (límite superior) & IndexOutOfBounds\\

\hline
\end{longtable}
}

\subsection{GameManager::connectToGame}

En este método tenemos dos parámetros, uno es el número de la partida seleccionada para jugar y el otro un evento, que para poder realizar la pruebas tenemos que crearnos una clase privada que implemente \textit{GameEventListener}. De esta forma, nos centramos en el primer parámetro que es el interesante. Como sólo tenemos un parámetro al igual que antes tenemos que sacar los valores límite. En esta ocasión en el servidor tenemos una única partida actual, por lo que el tope es ``0'' y el tope más uno es ``1''.

\begin{itemize}
\item \textbf{\texttt{gameSelected}}
\subitem Límite inferior: \texttt{"\--1"} y \texttt{"0"}
\subitem Límite superior: \texttt{"tope = 0"} y \texttt{"tope + 1 = 1"}
\end{itemize}

Además, antes de conectar tenemos que comprobar que almenos tengamos una partida a la cual podamos conectarnos para jugar y que el objeto GameEngine esté a null y que después de ejecutar el conectar no sea null.

Esta es la tabla de casos de prueba y resultados esperados.

{\footnotesize
\begin{longtable}[c]{lccc}
 & \textbf{Valores de Prueba} & \textbf{Objetivo del test} & \textbf{Resultado esperado} \\
\hline \hline
\endhead

Test1 & (-1, new TestGameEventListener()) & gameIndex (límite inferior) & ArrayIndexOutOfBounds\\
Test2 & (0, new TestGameEventListener()) & seleccionar partida & Funcionamiento correcto\\
Test3 & (1, new TestGameEventListener()) & gameIndex (límite superior) & IndexOutOfBounds\\
Test4 & (0, null) & gameListener & NullPointer\\

\hline
\end{longtable}
}

\subsection{GameManager::disconnectFromGame}

Este método no tiene ningún parámetro de entrada, por lo que para comprobar que funciona correctamente. Tenemos que conectarnos previamante a una partida y luego ejecutar este método, si se lanza la excepción \textit{NullPointerExcepcition}, la ejecución fallará. En caso contrario el método funciona correctamente.


